bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/main/java/com/google/cloud/bigtable/mapreduce/validation/BigtableSyncMapper.java

```diff
@@ -0,0 +1,187 @@
+    if (!conf.onlyKeyExists(TARGET_BT_PROJECTID_CONF_KEY)) {
+      createBigtableTargetConnection(conf);
+    }
+
+    // establish source connection to bigtable
+    if (!conf.onlyKeyExists(SOURCE_BT_PROJECTID_CONF_KEY)) {
+      createBigtableSourceConnection(conf);
+    }
+  }
+
+  /**
+   * create source connection with Bigtable
+   *
+   * @param conf
+   * @throws IOException
+   */
+  private void createBigtableSourceConnection(Configuration conf) throws IOException {
+    // re-create connection with bigtable
+    closeConnection(BigtableSyncTableAccessor.getSourceConnection(this), "source");
+
+    // create source connection config
+    // inherit base config, but override connection based on job args
+    Configuration bigtableConf =
+        getBigtableConfiguration(
+            new Configuration(conf),
+            SOURCE_BT_PROJECTID_CONF_KEY,
+            SOURCE_BT_INSTANCE_CONF_KEY,
+            SOURCE_BT_APP_PROFILE_CONF_KEY,
+            "source");
+    Connection srcConn = BigtableSyncTableAccessor.setSourceConnection(this, bigtableConf);
+    BigtableSyncTableAccessor.setSourceTable(this, srcConn, bigtableConf);
+  }
+
+  /**
+   * create target connection with Bigtable
+   *
+   * @param conf
+   * @throws IOException
+   */
+  private void createBigtableTargetConnection(Configuration conf) throws IOException {
+    // re-create connection with bigtable
+    closeConnection(BigtableSyncTableAccessor.getTargetConnection(this), "target");
+
+    // create target connection config
+    // inherit base config, but override connection based on job args
+    Configuration bigtableConf =
+        getBigtableConfiguration(
+            new Configuration(conf),
+            TARGET_BT_PROJECTID_CONF_KEY,
+            TARGET_BT_INSTANCE_CONF_KEY,
+            TARGET_BT_APP_PROFILE_CONF_KEY,
+            "target");
+    Connection targetConn = BigtableSyncTableAccessor.setTargetConnection(this, bigtableConf);
+    BigtableSyncTableAccessor.setTargetTable(this, targetConn, bigtableConf);
+  }
+
+  /**
+   * close connection
+   *
+   * @param conn
+   * @param sourceOrTarget
+   */
+  private void closeConnection(Connection conn, String sourceOrTarget) {
+    try {
+      if (conn != null) {
+        conn.close();
+      }
+    } catch (IOException ioe) {
+      LOG.debug("error closing temporary " + sourceOrTarget + " connection, " + ioe.getMessage());
+      ioe.printStackTrace();
+    }
+  }
+
+  /**
+   * create bigtable configuration
+   *
+   * @param conf
+   * @param projectIdKey
+   * @param instanceKey
+   * @param appProfileKey
+   * @param sourceOrTarget
+   * @return
+   */
+  private Configuration getBigtableConfiguration(
+      Configuration conf,
+      String projectIdKey,
+      String instanceKey,
+      String appProfileKey,
+      String sourceOrTarget) {
+    Configuration bigtableConf = new Configuration(conf);
+
+    bigtableConf.unset(ClusterConnection.HBASE_CLIENT_CONNECTION_IMPL);
```

Comment:
is unset necessary? won't .configure() overwrite it?

Reply:
good suggestion, removed this and verified that it's overwritten as mentioned.

Reply:
this should be closed as we've removed the unset for bigtableConf